text/0022-pibd-deployment.md

```diff
@@ -0,0 +1,444 @@
+
+#### Requesting Segments
+
+Once the required list of segments has been derived, the node can then begin to request segments in parallel from peers according to the [peer selection strategy](#peer-selection) above. Exactly how the node does this is left as an implementation detail, but the following guidelines should be kept in mind:
+
+* MMRs are an append-only structure, and can only be reconstucted sequentially left-to-right. Therefore there is little point requesting too many segments for a given MMR in advance, particularly if memory is limited.
+* A reasonable effort should be made to spread requests across the 3 UTXO MMRs to allow their reconstruction to complete around the same time.
+
+If a requested segment is not received in a pre-determined amount of time, the node may re-request the segment from another peer.
+
+The core implementation uses these values to govern its requesting of segments. Note these are arbitrarily chosen with the aims of keeping memory cache requirements low and avoiding requesting too many segments in advance of what the 'desegmenter' is ready for:
+
+```
+/// Maximum number of received segments to cache (across all MMRs) before we stop requesting others
+pub const MAX_CACHED_SEGMENTS: usize = 15;
+
+/// How long the state sync should wait after requesting a segment from a peer before
+/// deciding the segment isn't going to arrive. The syncer will then re-request the segment
+pub const SEGMENT_REQUEST_TIMEOUT_SECS: i64 = 60;
+
+/// Number of simultaneous requests for segments we should make. Note this is currently
+/// divisible by 3 to try and evenly spread requests amount the 3 main MMRs (Bitmap segments
+/// will always be requested first)
+pub const SEGMENT_REQUEST_COUNT: usize = 15;
+```
+
+#### Receiving and Validating Segments
+
+Upon receiving a requested segment, the node should immediately validate the segment's merkle proof as described in the [PIBD Messages RFC](https://github.com/mimblewimble/grin-rfcs/blob/master/text/0020-pibd-messages.md). Note that validation of kernels and rangeproofs are a separate concern, and can be done after all segments have been downloaded and all MMRs reconstructed.
+
+If the segment fails validation, the node should not immediately ban the peer without checking to see whether the [horizon header](#horizon-header-height) has changed. 
+
+Nodes should cache received and validated segments until their contents are ready to be applied sequentially to their respective MMRs.
+
+### Applying Segments to MMRs
+
+As described in the [PIBD Messages RFC](https://github.com/mimblewimble/grin-rfcs/blob/master/text/0020-pibd-messages.md), segment messages contain lists of hashes and leaves as well as their MMR positions. Hashes and leaves should be sequentially applied to their respective MMRs in order to rebuild them.
```

Comment:
For future reference, we could add a note here that any hashes not at the root of maximal spent subtrees could be optimized away in the future.

Reply:
Added note